api/src/main/java/io/opentelemetry/api/trace/TraceStateBuilder.java

```diff
@@ -50,20 +54,21 @@ public TraceStateBuilder set(String key, String value) {
       return this;
     }
     // Initially create the Entry to validate input.
-    TraceState.Entry entry = TraceState.Entry.create(key, value);
     if (entries == null) {
       // Copy entries from the parent.
       entries = new ArrayList<>(parent.getEntries());
     }
-    for (int i = 0; i < entries.size(); i++) {
-      if (entries.get(i).getKey().equals(entry.getKey())) {
+    for (int i = 0; i < entries.size(); i += 2) {
+      if (entries.get(i).equals(key)) {
+        entries.remove(i);
```

Comment:
This loop has a lot of shadiness :) Relies on` entries.size()` being invoked on every loop iteration (true in every language, but not something people expect to matter I think), and two calls to the same `entries.remove(i)` ouch. Let's use `ListIterator.remove` instead.

Reply:
Thought a bit more and realized the suggestion doesn't make much sense. Hmm - let's at least add comments if nothing comes to mind to clean up in this PR.

Reply:
It doesn't actually rely on entries.size() being called each time, since we break as soon as we remove one. I'll extract out the size and add some comments.